datafusion/src/physical_plan/sorts/sort_preserving_merge.rs

```diff
@@ -128,6 +128,12 @@ impl ExecutionPlan for SortPreservingMergeExec {
         Distribution::UnspecifiedDistribution
     }
 
+    fn should_repartition_children(&self) -> bool {
+        // if the children are repartitioned they may no longer remain
+        // sorted
+        false
```

Comment:
this is the fix, the rest is tests

Reply:
I made this comment elsewhere, but going to make it here for visibility - this will only prevent repartitioning of direct children. 

So for example if you had

```
Sort Preserving Merge <- Projection Exec <- Sorted Scan
```

I think Repartition could convert this to

```
Sort Preserving Merge <- Projection Exec <- Repartition <- Sorted Scan
```

Reply:
I will write some more tests and fix this